File: 프로그래머스/lv2/42842. 카펫/카펫.js
// 카펫의 가로 길이는 세로 길이와 같거나, 세로 길이보다 길다
// 제한사항: yellow는 1 이상 2,000,000 이하인 자연수입니다.


// 전체의 가로길이는? 노랑의 가로길이 + 2
// 전체의 세로 길이? 노랑의 세로길이 + 2
// 노랑 가로길이를 x, 세로 길이를 y라고 하자
// xy = yellow, (x+2)(y+2) = brown
  // 2x+2y+4 = brown
  // y = (brown-2x-4)/2
  // x*y = yellow
  // x(brown-2x-4)/2 = yellow 이 식 풀면, x,y나옴 -> 근의공식!

function solution(brown, yellow) {
  let [x,y] = [0,0]
  let b = (4-brown)/2 //근의 공식에서 'b'담당

  x = (-b+Math.sqrt(Math.pow(b,2)-4*yellow))/2
  y = yellow/x

  console.log(x,y)

  return [Math.max(x,y)+2,Math.min(x,y)+2]

}

console.log(solution(24,24))

File: 프로그래머스/2/42842. 카펫/카펫.py
# 카펫의 가로 길이는 세로 길이와 같거나, 세로 길이보다 길다
# 노란색 격자의 가로, 세로를 a, b라고 하면
    # a*b = yellow - (1)
    # (a+2)*(b+2) = brown+yellow
    # -> 2(a+b)+4 = brown - (2)
    # yellow의 약수쌍을 구하고, 이 중에서 (1), (2)를 만족하는 (a+2), (b+2)를 return

def solution(brown, yellow):
    divisor_pair =[]
    # 약수 구하기
    for k in range(1,int((yellow)**(1/2))+1):
        if (yellow) % k == 0:
            divisor_pair.append([k,(yellow)//k])

    for divisor in divisor_pair:
        if (divisor[0]+divisor[1]) == (brown-4)//2:
            return [max(divisor)+2, min(divisor)+2]

print(solution(24,24))
print(solution(4004,999999)) # 1003, 1001